File: app/src/main/java/org/schabi/newpipe/settings/notifications/NotificationsChannelsConfigFragment.java
package org.schabi.newpipe.settings.notifications;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import org.schabi.newpipe.R;
import org.schabi.newpipe.database.subscription.NotificationMode;
import org.schabi.newpipe.local.subscription.SubscriptionManager;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public final class NotificationsChannelsConfigFragment extends Fragment
        implements NotificationsConfigAdapter.ModeToggleListener {

    private NotificationsConfigAdapter adapter;
    @Nullable
    private Disposable loader = null;
    private CompositeDisposable updaters;

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        adapter = new NotificationsConfigAdapter(this);
        updaters = new CompositeDisposable();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater,
                             @Nullable final ViewGroup container,
                             @Nullable final Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_channels_notifications, container, false);
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final RecyclerView recyclerView = view.findViewById(R.id.recycler_view);
        recyclerView.setAdapter(adapter);
    }

    @Override
    public void onActivityCreated(@Nullable final Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (loader != null) {
            loader.dispose();
        }
        loader = new SubscriptionManager(requireContext())
                .subscriptions()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(adapter::update);
    }

    @Override
    public void onDestroy() {
        if (loader != null) {
            loader.dispose();
        }
        updaters.dispose();
        super.onDestroy();
    }

    @Override
    public void onModeToggle(final int position, @NotificationMode final int mode) {
        final NotificationsConfigAdapter.SubscriptionItem subscription = adapter.getItem(position);
        updaters.add(
                new SubscriptionManager(requireContext())
                        .updateNotificationMode(subscription.getServiceId(),
                                subscription.getUrl(), mode)
                        .subscribeOn(Schedulers.io())
                        .subscribe()
        );
    }
}

File: app/src/main/java/org/schabi/newpipe/settings/notifications/NotificationsChannelsConfigFragment.kt
package org.schabi.newpipe.settings.notifications

import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers
import org.schabi.newpipe.R
import org.schabi.newpipe.database.subscription.NotificationMode
import org.schabi.newpipe.local.subscription.SubscriptionManager
import org.schabi.newpipe.settings.notifications.NotificationsConfigAdapter.ModeToggleListener

class NotificationsChannelsConfigFragment : Fragment(), ModeToggleListener {

    private lateinit var updaters: CompositeDisposable
    private var loader: Disposable? = null
    private var adapter: NotificationsConfigAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        updaters = CompositeDisposable()
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View = inflater.inflate(R.layout.fragment_channels_notifications, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val recyclerView: RecyclerView = view.findViewById(R.id.recycler_view)
        adapter = NotificationsConfigAdapter(this)
        recyclerView.adapter = adapter
        loader?.dispose()
        loader = SubscriptionManager(requireContext())
            .subscriptions()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { newData -> adapter?.update(newData) }
    }

    override fun onDestroyView() {
        loader?.dispose()
        loader = null
        super.onDestroyView()
    }

    override fun onDestroy() {
        updaters.dispose()
        super.onDestroy()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_notifications_channels, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_toggle_all -> {
                toggleAll()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onModeToggle(position: Int, @NotificationMode mode: Int) {
        val subscription = adapter?.getItem(position) ?: return
        updaters.add(
            SubscriptionManager(requireContext())
                .updateNotificationMode(
                    subscription.serviceId,
                    subscription.url,
                    mode
                )
                .subscribeOn(Schedulers.io())
                .subscribe()
        )
    }

    private fun toggleAll() {
        val subscriptions = adapter?.getCurrentList() ?: return
        val mode = subscriptions.firstOrNull()?.notificationMode ?: return
        val newMode = when (mode) {
            NotificationMode.DISABLED -> NotificationMode.ENABLED
            else -> NotificationMode.DISABLED
        }
        val subscriptionManager = SubscriptionManager(requireContext())
        updaters.add(
            CompositeDisposable(
                subscriptions.map { item ->
                    subscriptionManager.updateNotificationMode(
                        serviceId = item.serviceId,
                        url = item.url,
                        mode = newMode
                    ).subscribeOn(Schedulers.io())
                        .subscribe()
                }
            )
        )
    }
}
